src/routing_node.rs

```diff
@@ -1138,35 +1154,139 @@ impl RoutingNode {
+                                        }
+                                    }
+                                }
+                            }
+                            _ => {}
                         }
-                    }
-                    _ => None
+
+                    },
+                    _ => {}
                 }
-            }
-            _ => None,
+            },
+            None => {}
         }
+            
     }
 
-    fn handle_cache_put(&mut self, response: ExternalResponse) {
-        match response {
-            ExternalResponse::Get(data, _, _) => {
-                match data {
-                    Data::ImmutableData(ref immutable_data) => {
-                        self.data_cache.insert(immutable_data.name(), data.clone());
-                    }
-                    _ => {}
+    fn handle_cache_get(&mut self, message: &RoutingMessage) -> Option<Content> {
+        match self.data_cache {
+            Some(ref mut data_cache) => {
+                match message.content.clone() {
+                    Content::ExternalRequest(request) => {
+                        match request {
+                            ExternalRequest::Get(data_request, _) => {
+                                match data_request {
+                                    DataRequest::PlainData(data_name) => {
+                                        if self.cache_options.plain_data_caching_enabled() {
+                                            match data_cache.get(&data_name) {
+                                                Some(data) => {
+                                                    let response =
+                                                        ExternalResponse::Get(
+                                                            data.clone(),
+                                                            data_request,
+                                                            None);
+                                                    return Some(Content::ExternalResponse(response));
+                                                },
+                                                None => return None
+                                            }
+                                        }
+                                        return None;
+                                    }
+                                    DataRequest::StructuredData(data_name, version) => {
+                                        if self.cache_options.structured_data_caching_enabled() {
+                                            match data_cache.get(&data_name) {
+                                                Some(data) => {
+                                                    match *data {
+                                                        Data::StructuredData(ref structured_data) => {
+                                                            if structured_data.get_version() == version {
```

Comment:
So here and on line 1239 the name of the structured data needs to be calculated and used - no version number checking required.  The data_cache::get (and corresponding put) needs to follow

DataRequest::StructuredData(ref name, tag) => StructuredData::compute_name(tag, name)